src/command.rs

```diff
@@ -424,34 +425,533 @@ impl <U:WebDriverExtensionRoute> ToJson for WebDriverMessage<U> {
+                    ErrorStatus::InvalidArgument,
+                    format!("{} was not a valid proxy configuration capability", x)))
+            }
+        }
+        Ok(())
+    }
+
+    fn validate_host_domain(&self,
+                            name: &str,
+                            scheme: &str,
+                            obj: &Capabilities,
+                            value: &Json) -> WebDriverResult<()> {
+        match value.as_string() {
+            Some(x) => {
+                if x.contains("::/") {
+                    return Err(WebDriverError::new(
+                        ErrorStatus::InvalidArgument,
+                        format!("{} contains a scheme", name)))
+                };
+                let mut s = String::with_capacity(scheme.len() + x.len() + 3);
+                s.push_str(scheme);
+                s.push_str("://");
+                s.push_str(x);
+                let url = try!(Url::parse(&*s).or(Err(WebDriverError::new(
+                    ErrorStatus::InvalidArgument,
+                    format!("{} was not a valid url", name)))));
+                if url.username() != "" ||
+                    url.password() != None ||
+                    url.path() != "/" ||
+                    url.query() != None ||
+                    url.fragment() != None {
+                        return Err(WebDriverError::new(
+                            ErrorStatus::InvalidArgument,
+                            format!("{} was not of the form host[:port]", name)));
+                    }
+                let mut port_key = String::with_capacity(name.len() + 4);
+                port_key.push_str(name);
+                port_key.push_str("Port");
+                if url.port() != None &&
+                    obj.contains_key(&*port_key) {
+                        return Err(WebDriverError::new(
+                                    ErrorStatus::InvalidArgument,
+                                    format!("{} supplied with a port as well as {}",
+                                            name, port_key)));
+                    }
+            },
+            None => return Err(WebDriverError::new(
+                ErrorStatus::InvalidArgument,
+                format!("{} was not a string", name)
+            ))
+        }
+        Ok(())
+    }
+
+    fn validate_port(&self, name: &str, value: &Json) -> WebDriverResult<()> {
+        match value {
+            &Json::I64(x) => {
+                if x < 0 || x > 2^16 - 1 {
+                    return Err(WebDriverError::new(
+                        ErrorStatus::InvalidArgument,
+                        format!("{} is out of range", name)))
+                }
+            }
+            _ => return Err(WebDriverError::new(
+                ErrorStatus::InvalidArgument,
+                format!("{} was not an integer", name)))
+        }
+        Ok(())
+    }
+
+    fn validate_timeouts(&self, value: &Json) -> WebDriverResult<()> {
```

Comment:
This is (almost) duplicated in https://github.com/mozilla/webdriver-rust/blob/master/src/command.rs#L530. It would be nice to reuse that, but don’t let that block this PR.